### PR TITLE
Support more sugared extension points

### LINKED ISSUE #410: Recognise all sugars for extension points.
Also, consider always preserving the syntax of extension points, rewriting between syntaxes can be done outside ocamlformat.

### LINKED ISSUE #1584: Feature request: Option to prevent transform `assert%lwt false` to `[%lwt assert false]`
I've tried both `extension-sugar={preserve|always}`. But ocamlformat always transform my `assert%lwt false` to `[%lwt assert false]`

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #1584 and fix #410, no diff with test_branch.

~I will probably remove the option `extension-sugar` that is deprecated since 0.14.0 first, this will reduce the diff of this PR => #1588~

### CODE REVIEW

**Comment by hhugo** [comment]
Does this add support all sugars ?

**Comment by gpetiot** [comment]
> Does this add support all sugars ?

I noticed the support was missing for signature items, I'm working on it and it should be alright when I'll undraft this PR.

**Comment by hhugo** [comment]
`begin%ext ... end` is most probably missing but it is another kind of beast and should not be addressed here.

**Comment by gpetiot** [comment]
I think every case is handled now judging from the source.ml test file, I will have a look at the `begin%ext ... end` in another PR. No diff with test branch on the conventional and janestreet profile.

**Comment by hhugo** [comment]
> I will have a look at the `begin%ext ... end` in another PR

Can you add a test for it now ?

**Comment by emillon** [comment]
There's `;%ext` too, not sure if it's already supported or not.

**Comment by gpetiot** [comment]
> There's `;%ext` too, not sure if it's already supported or not.

It is!

**Comment by hhugo** [comment]
What about the expression `(module%ext S : S)` ?

**Comment by gpetiot** [comment]
> What about the expression `(module%ext S : S)` ?

It is handled now

**Comment by hhugo** [comment]
Have you check in the ocaml parser that all sugar cases are handled by ocamlformat ? (except begin%ext)

File: lib/Ast.ml [hunk 1]
```diff
@@ -1742,28 +1742,9 @@ end = struct
...
-                ; _ } ] )
-        when Source.extension_using_sugar ~name:ext ~payload:e.pexp_loc ->
-          Some Apply
-      | Pexp_extension
-          ( ext
-          , PStr
-              [ { pstr_desc=
-                    Pstr_eval (({pexp_desc= Pexp_sequence _; _} as e), _)
-                ; _ } ] )
+      | Pexp_extension (ext, PStr [{pstr_desc= Pstr_eval (e, _); _}])
```
Comment (hhugo) [inline lib/Ast.ml:1745]: can you explain this change ?

File: lib/Ast.ml [hunk 2]
```diff
@@ -1742,28 +1742,9 @@ end = struct
...
         when Source.extension_using_sugar ~name:ext ~payload:e.pexp_loc ->
-          Some Semi
+          prec_ast (Exp e)
```
Comment (hhugo) [inline lib/Ast.ml:1747]: can you explain this change ?

File: lib/Ast.ml (hunk 1 above)
Comment (gpetiot) [reply lib/Ast.ml:1745]: Because in the case where we use the sugared syntax this function should return the same for `keyword%ext` and `keyword`

**Comment by gpetiot** [comment]
> Have you check in the ocaml parser that all sugar cases are handled by ocamlformat ? (except begin%ext)

I had a look at the ocaml parser and I think all cases are handled now

**Review by hhugo — APPROVED** [review]

**Pull request merged**